README.md
# Outbrain React Native Demo

Project used [this tutorial](https://www.freecodecamp.org/news/how-to-publish-a-react-native-component-to-npm-its-easier-than-you-think-51f6ae1ef850/) for referene.

use `npm pack` to test locally

See the [Release Notes](release-notes.md) for the latest version details.

## Installation

Using npm:

```shell
npm install --save react-native-outbrain
```

or using yarn:

```shell
yarn add react-native-outbrain
```

## Widget Settings (AMPM)

### Custom Clicks Handler (982) = enable
Must be set so all click will be passed to the native function. if a feed is in use then change this setting to all the feed child widgets

### Feed Manual Chunk Fetch (919) = enable
On ReactNative the native scroll will not work, managing the fetching of next cards is done from the app.

In general please refer to this [Outbrain Confluence page](https://confluence.outbrain.com/display/CATS/React+Native+Bridge+project)


## Example Apps - Download Links

[iOS Demo download link](https://install.appcenter.ms/users/oregev/apps/react-native-ios-demo/distribution_groups/public)

[Android Demo download link](https://install.appcenter.ms/users/oregev/apps/react-native-android-demo/distribution_groups/public)


## Usage

### Import Outbrain Module

```
import { OutbrainWidget } from 'react-native-outbrain'
```

### Example for displaying a regular widget

```
<OutbrainWidget
  url={"http://mobile-demo.outbrain.com"}
  widgetId={"MB_2"}
  installationKey={"NANOWDGT01"}
  userId={this.state.IDFA}
/>
```

### Example for displaying a Smartfeed widget

```
<OutbrainWidget
  url={"http://mobile-demo.outbrain.com"}
  widgetId={"MB_1"}
  installationKey={"NANOWDGT01"}
  ref={input => this.outbrainWidget = input}
  userId={this.state.IDFA}
/>
```

### Example for detect scroll to bottom

```
isCloseToBottom ({layoutMeasurement, contentOffset, contentSize}) {
  const paddingToBottom = 50;
  return layoutMeasurement.height + contentOffset.y >= contentSize.height - paddingToBottom;
};

<ScrollView
  style={styles.container}
  showsVerticalScrollIndicator={false}
  onMomentumScrollEnd={({nativeEvent}) => {
    if (this.outbrainWidget && this.isCloseToBottom(nativeEvent)) {
      console.log("onMomentumScrollEnd - isCloseToBottom")
      this.outbrainWidget.loadMore()
    }
  }}>
```

### GDPR \ CCPA Support

Use the following props to pass `OutbrainWidget` the GDPR or CCPA string from your app code.

For GDPR consent V1 use `consentV1`
For GDPR consent V2 use `consentV2`
For CCPA string use `ccpaString`


```
<OutbrainWidget
                url={"http://mobile-demo.outbrain.com"}
                widgetId={"MB_1"}
                installationKey={"NANOWDGT01"}
                consentV1={"1111111"}
                consentV2={"2222222"}
                ccpaString={"3333333"}
                style={{width: screenWidth*1}}
                ref={input => this.outbrainWidget = input}
                userId={this.state.IDFA}
              />
```

### Example for getting IDFA

```
import { IDFA } from 'react-native-idfa';

componentDidMount() {
  IDFA.getIDFA().then((idfa) => {
    this.setState({ IDFA: idfa, });
  })
  .catch((e) => {
    console.error(e);
  });
}
```

### Example for organic click custom listener

```
<OutbrainWidget
  url={"http://mobile-demo.outbrain.com"}
  widgetId={"MB_1"}
  installationKey={"NANOWDGT01"}
  onOrganicClick={orgUrl => console.log("In App - click on: " + orgUrl)}
  style={{width: screenWidth*1}}
  ref={input => this.outbrainWidget = input}
  userId={this.state.IDFA}
/>
```

### Dark mode support

To enable dark mode, set `darkMode` in props to true.

```
<OutbrainWidget
  url={"http://mobile-demo.outbrain.com"}
  widgetId={"MB_1"}
  installationKey={"NANOWDGT01"}
  onOrganicClick={orgUrl => console.log("In App - click on: " + orgUrl)}
  style={{width: screenWidth*1}}
  ref={input => this.outbrainWidget = input}
  userId={this.state.IDFA}
  darkMode={true}
/>
```

### Development Setup

Edit package.json with `"react-native-outbrain": "../../"`

### Test a new version

1. update version in package.json
2. run `npm pack` --> creates `eact-native-outbrain-1.0.3.tgz` for example.
3. in /examples/examples/DemoProject `yarn add ../../react-native-outbrain-1.0.3.tgz`

### Release a new version

1. Edit version in package.json in root dir
2. npm login
3. npm publish


release-notes.md
# outbrain-react-native module - Release Notes

## v1.4.0 - Jan 30, 2022

- Fix crash on Android 11+ with react-native-webview

## v1.3.1 - Jan 25, 2022

- Dark mode support - new prop for `OutbrainWidget` - `darkMode`.
- Change default background of `OutbrainWidget` to transparent.

## v1.3.0 - June 14, 2021

- Added a new prop for `OutbrainWidget` - `webViewHeightPadding` following a request from publisher.
- Fixed issue with flipper and Xcode 12.5 - [see reference](https://stackoverflow.com/questions/66189325/xcode-throws-atomic-notify-oneunsigned-long-is-unavailable)


## v1.2.1 - Feb 8, 2021

- Fix: Video autoplay on iOS opens in full screen instead of inline.

## v1.2.0 - Dec 16, 2020

- Feature: support GDPR, CCPA string - pass as props to `OutbrainWidget`
- Fix dependency react-native-webview error if included twice in a project [#8](https://github.com/outbrain/outbrain-react-native-component/pull/8)
- Bump react-native-webview from 10.10.0 to 11.0.0

## v1.1.0 - Nov 4, 2020

- Infrastructure - update to React Native 0.63.3 (was 0.60.0)
- Infrastructure - update to iOS14
- Update lib "react-native-webview": "^10.10.0"

## v1.0.3 - Jul 7, 2020

- Feature - add support for onOrganicClick()

## v1.0.2 - Apr 5, 2020

- Feature - add support for installationKey param

## v1.0.1 - Feb 10, 2020

- First release


examples/DemoProject/app/ArticleScreen.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */

import PropTypes from 'prop-types'
import React, {Fragment, Component} from 'react';
import {
  SafeAreaView,
  StyleSheet,
  ScrollView,
  Alert,
  View,
  Image,
  Text,
  Linking,
  Dimensions,
  StatusBar,
} from 'react-native';

import { IDFA } from 'react-native-idfa';

import { OutbrainWidget } from 'react-native-outbrain'


export default class ArticleScreen extends Component {

  constructor (props) {
    super(props)
    this.state = {
      showSmartfeed: props.route.params.isSmartfeed,
      showRegularWidget: props.route.params.isRegularWidget,
      deviceIDFA: null,
    }
  }

  componentDidMount() {
    IDFA.getIDFA().then((idfa) => {
      this.setState({ deviceIDFA: idfa.length > 0 ? idfa : null });
    })
    .catch((e) => {
      console.error(e);
    });
  }

  isCloseToBottom ({layoutMeasurement, contentOffset, contentSize}) {
    const paddingToBottom = 50;
    return layoutMeasurement.height + contentOffset.y >= contentSize.height - paddingToBottom;
  };

  render() {
    const screenWidth = Math.round(Dimensions.get('window').width);
    return (
      <Fragment>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView style={{flex: 1}}>
          <ScrollView
            showsVerticalScrollIndicator={false}
            onMomentumScrollEnd={({nativeEvent}) => {
              if (this.outbrainWidget && this.isCloseToBottom(nativeEvent)) {
                console.log("onMomentumScrollEnd - isCloseToBottom")
                this.outbrainWidget.loadMore()
              }
            }}>
            <View style={styles.container}>

              <Text style={styles.articleTitle}>
                {"Jose Mourinho agrees deal to replace Mauricio Pochettino as Tottenham manager"}
              </Text>
              <Text style={styles.articleSubtitle}>
                {"The Portuguese boss had been out of work since being sacked by Manchester United in December 2018."}
              </Text>
              <View style={styles.articleAuthorContainer}>
                <Image
                  style={{width: 65, height: 65, borderRadius: 32}}
                  source={require('./assets/Images/skynews-tom-gillespie-news-reporter_4616217.jpg')} />
                  <View style={{marginLeft: 15, marginTop: 5, alignSelf: 'flex-start'}}>
                    <Text style={styles.authorName}>
                      {"Tom Gillepie"}
                    </Text>
                    <Text style={styles.authorDescription}>
                      {"News reporter @TomGillespie1"}
                    </Text>
                  </View>
              </View>
              <Text style={styles.articleDatetime}>
                {"Wednesday 20 November 2019 12:12, UK"}
              </Text>

              <Image
                style={{height: 250, marginLeft: -20, width: screenWidth, marginTop: 10, resizeMode: 'contain'}}
                source={require('./assets/Images/skynews-jose-mourinho_4841610.jpg')} />

              <Text style={styles.articleContent}>
                {"Jose Mourinho has agreed a deal to replace Mauricio Pochettino as Tottenham's manager until the end of the 2022/23 season.\
                \n\nPochettino was sacked last night after five-and-a-half years in charge, and less than six months after leading Spurs to their\
                 first ever Champions League final.\
                 \n\nNegotiations between the London side and Mourinho's representatives intensified over the last few days, with talks continuing into Tuesday night.\
                 \n\nThe Portuguese boss had been out of work since being sacked by Manchester United in December 2018.\
                 \n\nThe 56-year-old said after his appointment: \"I am excited to be joining a club with such a great heritage and such passionate supporters."}
              </Text>

              {this.state.showRegularWidget &&
                <OutbrainWidget
                  url={"http://mobile-demo.outbrain.com"}
                  widgetId={"MB_2"}
                  installationKey={"NANOWDGT01"}
                  userId={this.state.IDFA}
                  style={{ width: screenWidth*0.9}}
                />
              }

              <Text style={styles.articleContent}>
                {"Jose Mourinho has agreed a deal to replace Mauricio Pochettino as Tottenham's manager until the end of the 2022/23 season.\
                \n\nPochettino was sacked last night after five-and-a-half years in charge, and less than six months after leading Spurs to their\
                 first ever Champions League final.\
                 \n\nNegotiations between the London side and Mourinho's representatives intensified over the last few days, with talks continuing into Tuesday night.\
                 \n\nThe Portuguese boss had been out of work since being sacked by Manchester United in December 2018.\
                 \n\nThe 56-year-old said after his appointment: \"I am excited to be joining a club with such a great heritage and such passionate supporters."}
              </Text>
            </View>

            {this.state.showSmartfeed &&
              <OutbrainWidget
                url={"http://mobile-demo.outbrain.com"}
                widgetId={"MB_1"}
                installationKey={"NANOWDGT01"}
                style={{width: screenWidth*1}}
                ref={input => this.outbrainWidget = input}
                userId={this.state.IDFA}
              />
            }

          </ScrollView>
        </SafeAreaView>
      </Fragment>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    paddingLeft: 20,
    paddingRight: 20,
  },
  articleTitle: {
    fontSize: 30,
    fontWeight: '300'
  },
  articleSubtitle: {
    marginTop: 20,
    fontSize: 18,
    fontWeight: '200'
  },
  articleAuthorContainer: {
    marginTop: 15,
    flexDirection: 'row',
    alignItems: 'center',
    alignSelf: 'flex-start'
  },
  authorName: {
    fontSize: 14,
    color: "#063e7b"
  },
  authorDescription: {
    marginTop: 5,
    fontSize: 14,
    color: "#707070"
  },
  articleDatetime: {
    marginTop: 20,
    fontSize: 14,
    color: '#9f9f9f',
    alignSelf: 'flex-start'
  },
  articleContent: {
    marginTop: 25,
    fontSize: 16,
    color: '#4a4a4a'
  }
});
